### PR TITLE
Bug spot

### LINKED ISSUE #174: variable issue reported by spotbugs
**Describe the bug**
After running the tool called spotbugs, we get below warning message:

- Dead store to the local variable
eg. Variable “pgLines” in org.qcmg.qprofiler2.bam.BamSummarizer2::createReport is stored value but not used,  it should be deleted. 

-  Unread public/protected field
eg. org.qcmg.qprofiler2.summarise.PairSummary.logger is not used and should be delete.

- Unread field
eg. field: org.qcmg.qprofiler2.summarise.SampleSummary.sampleTrans, is not read and should be deleted.

- Field isn't final but should be
eg. org.qcmg.qprofiler2.cohort.CohortSummaryReport.headerline isn't final but should be.

- Boxed value is unboxed and then immediately reboxed
eg org.qcmg.qprofiler2.util.XmlUtils.outputValueNode(Element parent, String name, Number value)
`{ String v = String.valueOf(value); … v = String.format("%,.2f", (double)value); } ` 
should use auto box and change to
`{ String v = value + “”; … v = String.format("%,.2f", value);`

-  Method uses the same code for two branches
eg. new org.qcmg.qprofiler2.summarise.CycleSummary(Object, int) uses the same code for two branches, the code was
```
int DEFAULT_NO_OF_KEYS_CHARACTER = 128;
int DEFAULT_NO_OF_KEYS_INTEGER = 128;
int noOfKeys = type instanceof Character? DEFAULT_NO_OF_KEYS_CHARACTER : DEFAULT_NO_OF_KEYS_INTEGER;
```
 It should change to one line
`int noOfKeys = 128;`

**To Reproduce**
Above issues are resolved in the comming pull request.

### PR DESCRIPTION
**Pull request by ChristinaXu2017**
# Description

run spotbugs on qcommon, most related bugs are fixed.  Fixes #174 #175 #176 #177 #178. However some problems are ignored,  such as missing default cases in switch statement, call start() method in threads. etc
 
## Type of change

Please delete options that are not relevant.

- [X ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

used existing unit test code

# Checklist:

- [X ] My code follows the style guidelines of this project
- [X ] I have performed a self-review of my own code
- [ X] I have commented my code, particularly in hard-to-understand areas
- [ X] I have made corresponding changes to the documentation
- [X ] My changes generate no new warnings
- [X ] I have added tests that prove my fix is effective or that my feature works
- [ X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by holmeso** [comment]
Could you pls use StringBuilder rather than StringBuffer - its quicker (doesn't synchronise)

**Comment by holmeso** [comment]
Why the move from static methods to instance methods in FisherExact class?

File: qcommon/src/org/qcmg/common/model/PileupElement.java [hunk 1]
```diff
@@ -73,12 +76,30 @@ public int getTotalCount() {
...
 	}
-
+	@Override
 	public int compareTo(PileupElement o) {
 		// only interested in the total count for the purposes or ordering
 		return o.getTotalCount() - getTotalCount();
 	}
 	
+	//you need to implement the equals method thus overriding the default compareTo
+	@Override
```
Comment (holmeso) [inline qcommon/src/org/qcmg/common/model/PileupElement.java:86]: this equals and hashcode methods look a little sparse.
Eclipse provides a decent implementation of these methods which will be a little more robust.
Right click in the class, go to `Source` -> `Generate hashCode and equals()`

File: qcommon/src/org/qcmg/common/model/PileupElementLite.java [hunk 2]
```diff
@@ -118,5 +118,24 @@ public int compareTo(PileupElementLite o) {
 		// only interested in the total count for the purposes or ordering
 		return o.getTotalCount() - getTotalCount();
 	}
+	
+	//you need to implement the equals method thus overriding the default compareTo
+	@Override
```
Comment (holmeso) [inline qcommon/src/org/qcmg/common/model/PileupElementLite.java:123]: need a stronger equals and hashCode

File: qcommon/src/org/qcmg/common/util/AccumulatorUtils.java [hunk 3]
```diff
@@ -464,6 +465,33 @@ else if (isRef)
 				return -1;
 			return o.quality - quality;
 		}
+		
+		//you need to implement the equals method thus overriding the default compareTo
+		@Override 
+		public boolean equals(Object o) {
```
Comment (holmeso) [inline qcommon/src/org/qcmg/common/util/AccumulatorUtils.java:471]: don't think we need equals and hashcode for a utility class with only static methods

File: qcommon/src/org/qcmg/common/util/LoadReferencedClasses.java [hunk 4]
```diff
@@ -20,8 +20,8 @@
 	
 	
 	public static final <T> void loadClasses(final Class<T> clazz) throws URISyntaxException, IOException, ClassNotFoundException {
-		final long start = System.currentTimeMillis();
-		final long mem = Runtime.getRuntime().freeMemory();
+//		final long start = System.currentTimeMillis();
```
Comment (holmeso) [inline qcommon/src/org/qcmg/common/util/LoadReferencedClasses.java:23]: delete

File: qcommon/src/org/qcmg/common/model/PileupElement.java (hunk 1 above)
Comment (ChristinaXu2017) [reply qcommon/src/org/qcmg/common/model/PileupElement.java]: thanks for suggestion

File: qcommon/src/org/qcmg/common/model/PileupElementLite.java (hunk 2 above)
Comment (ChristinaXu2017) [reply qcommon/src/org/qcmg/common/model/PileupElementLite.java]: replaced by eclipse provided one

File: qcommon/src/org/qcmg/common/util/AccumulatorUtils.java (hunk 3 above)
Comment (ChristinaXu2017) [reply qcommon/src/org/qcmg/common/util/AccumulatorUtils.java]: I agree, and removed

File: qcommon/src/org/qcmg/common/util/LoadReferencedClasses.java (hunk 4 above)
Comment (ChristinaXu2017) [reply qcommon/src/org/qcmg/common/util/LoadReferencedClasses.java]: tidied up

**Comment by holmeso** [comment]
FisherExact should really be deprecated and then removed in a future release

**Pull request merged**